pkg/proxy/request.go

```diff
@@ -0,0 +1,113 @@
+package proxy
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"strconv"
+
+	"github.com/google/uuid"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// proxyRequest keeps track of the state of a request that is to be proxied.
+type proxyRequest struct {
+	// r is the original request. It is stored here so that is can be cloned
+	// and sent to the backend while preserving the original request body.
+	r *http.Request
+	// body will be stored here if the request body needed to be read
+	// in order to determine the model.
+	body []byte
+
+	// metadata:
+
+	id                string
+	status            int
+	model             string
+	backendDeployment string
+	attempt           int
+
+	// metrics:
+
+	timer *prometheus.Timer
+}
+
+func newProxyRequest(r *http.Request) *proxyRequest {
+	pr := &proxyRequest{
+		r:      r,
+		id:     uuid.New().String(),
+		status: http.StatusOK,
+	}
+
+	pr.timer = prometheus.NewTimer(prometheus.ObserverFunc(func(v float64) {
+		httpDuration.WithLabelValues(pr.model, strconv.Itoa(pr.status)).Observe(v)
+	}))
+
+	return pr
+
+}
+
+func (p *proxyRequest) done() {
+	p.timer.ObserveDuration()
+}
+
+func (pr *proxyRequest) parseModel() error {
+	pr.model = pr.r.Header.Get("X-Model")
+	if pr.model != "" {
+		return nil
+	}
+
+	var err error
+	pr.body, err = io.ReadAll(pr.r.Body)
+	if err != nil {
+		return fmt.Errorf("read: %w", err)
+	}
+
+	var payload struct {
+		Model string `json:"model"`
+	}
+	if err := json.Unmarshal(pr.body, &payload); err != nil {
+		return fmt.Errorf("unmarshal json: %w", err)
+	}
+	pr.model = payload.Model
+
+	if pr.model == "" {
+		return fmt.Errorf("no model specified")
+	}
+
+	return nil
+}
+
+func (pr *proxyRequest) sendErrorResponse(w http.ResponseWriter, status int, format string, args ...interface{}) {
```

Comment:
A helper function for writing errors makes a lot of sense. I had this in my head for some time already. ;-)
In the implementation you don't differentiate between log and the error message returned. A log is usually more verbose and may contain internals that you don't want to share with anonymous clients.

Reply:
I was tied between this and splitting the logging out completely. What I landed on was the `if status >= 500` check below that will filter out any messages on internal errors.

Reply:
Would like to get your take on this approach.

Reply:
I missed the condition, sorry. Let's give it a try